---
title: 19.并行执行异步函数 - 中等
---
给定一个异步函数数组 `functions`，返回一个新的 promise 对象 `promise`。数组中的每个函数都不接受参数并返回一个 promise。所有的 promise 都应该并行执行。

`promise` resolve 条件：

- 当所有从 `functions` 返回的 promise 都成功的并行解析时。`promise` 的解析值应该是一个按照它们在 `functions` 中的顺序排列的 promise 的解析值数组。`promise` 应该在数组中的所有异步函数并行执行完成时解析。

`promise` reject 条件：

- 当任何从 `functions` 返回的 promise 被拒绝时。`promise` 也会被拒绝，并返回第一个拒绝的原因。

请在不使用内置的 `Promise.all` 函数的情况下解决。

**示例 1：**

**输入：** 
```js
functions = [
  () => new Promise(resolve => setTimeout(() => resolve(5), 200))
]
```
**输出：** `{"t": 200, "resolved": [5]}`
**解释：**
`promiseAll(functions).then(console.log); // [5]`

单个函数在 200 毫秒后以值 5 成功解析。

解析：
模拟一个`promise.all()`功能
用`.then()` `.catch()`
```js
var promiseAll = function(functions){
	return new Promise((resolve, rejct) =>{
		if(functions.length === 0){
			resolve([]);
			return
		}
		
		const result = new Array(functions.length).fill(null);
		let resolvedCount = 0;
		
		functions.forEach((func, idx) =>{
			func().then((res)=>{
				result[idx] = res;
				resolvedCount ++;
				if(resolvedCount === functions.length){
					resolve(result)
				}
			}).catch((error)=>{
				reject(error)
			})
		})
	})
}
```
用`async/await`
```js
var promiseAll = function(functions){
	return new Promise((resolve, reject)=>{
		if(functions.length === 0){
			resolve([]);
			return
		}
		const result = new Array(functions.length).fill(null);
		let resolvedCount = 0;
		functions.forEach( async (func,idx)=>{
			try{
				const res = await func();
				result[idx] = res;
				resolvedCount++;
				if(resolvedCount === functions.length){
					resolve(result)
				}
			}catch(error){
				reject(error)
			}
		})
	})

}
```


### 面试提示：

1. Promise.all() 是什么，它是如何工作的？

Promise.all() 是 JavaScript 中的一个实用函数，它将多个 Promise 聚合成一个单一的 Promise，该 Promise 仅在所有输入 Promise 都已解决时才会解决，或者在输入 Promise 中的任何一个拒绝时立即拒绝。它通常用于需要同时执行多个异步操作，并且进一步的计算取决于这些操作的完成。

2. 如果传递给 Promise.all() 的 Promise 中有一个拒绝会发生什么？

如果传递给 Promise.all() 的 Promise 中有一个拒绝，Promise.all() 返回的 Promise 将立即被拒绝，并携带第一个拒绝的 Promise 的原因。这种行为有时被称为“快速失败”。

3. 如何处理 Promise.all() 中的单个 Promise 拒绝？

要处理 Promise.all() 中的单个 Promise 拒绝，你可以捕获单个 Promise 中的错误并将其转换为带有错误值的解决。这样，Promise.all() 将始终解决，而错误处理可以在生成的值数组上执行。但是，从 ECMAScript 2020 开始，更好的选择是使用 Promise.allSettled()。

4. Promise.all() 和 Promise.allSettled() 之间有什么区别？

Promise.allSettled() 方法与 Promise.all() 类似，但有一个关键区别。虽然 Promise.all() 只要其中一个 Promise 拒绝就会拒绝，Promise.allSettled() 在所有 Promise 已解决时或已拒绝时都会解决。Promise.allSettled() 的解析值是一个对象数组，每个对象都描述每个 Promise 的结果。

